### PR TITLE
Installing csi-proxy in some windows clusters

### PR DESCRIPTION
**Pull request by marosset**
Signed-off-by: Mark Rossetti <marosset@microsoft.com>

 

 
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
Installs csi-proxy as a daemonset on Windows nodes for clusters that specify `csi-proxy: enabled`
```

### CODE REVIEW

File: templates/addons/windows/csi-proxy/csi-proxy.yaml
Comment (jsturtevant) [inline templates/addons/windows/csi-proxy/csi-proxy.yaml:27]: newline

File: templates/test/ci/patches/windows-csi-proxy-enabled.yaml
Comment (jsturtevant) [inline templates/test/ci/patches/windows-csi-proxy-enabled.yaml:7]: newline

**Comment by jsturtevant** [comment]
private cluster timed out.  
/retest

**Review by jsturtevant — COMMENTED** [review]
Is the plan to follow up with a test config after? 

I am not 100% sure as I haven't tried this yet but I think it needs to be supported in a [few different templates](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/d41727fe7786b4ff4ba8618b5a353780826bb124/scripts/ci-entrypoint.sh#L67) including this one:

 https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/d41727fe7786b4ff4ba8618b5a353780826bb124/scripts/ci-entrypoint.sh#L89

**Comment by marosset** [comment]
> Is the plan to follow up with a test config after?

Yep

**Comment by marosset** [comment]
/test pull-kubernetes-e2e-capz-azure-disk-windows

It looks like some test jobs for windows + capz + azure-disk were added with https://github.com/kubernetes/test-infra/pull/24065

**Comment by marosset** [comment]
> > I am not 100% sure as I haven't tried this yet but I think it needs to be supported in a [few different templates](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/d41727fe7786b4ff4ba8618b5a353780826bb124/scripts/ci-entrypoint.sh#L67) including this one:
> 
> Thanks, I wasn't sure which all clusters this needed to be enabled for. Hopefully it will be clearer once I start the test-infra/config changes.

Nevermind, wrong repo :)

> /test pull-kubernetes-e2e-capz-azure-disk-windows
> 
> It looks like some test jobs for windows + capz + azure-disk were added with [kubernetes/test-infra#24065](https://github.com/kubernetes/test-infra/pull/24065)

Nevermind, wrong repo :)

**Comment by marosset** [comment]
/hold
While I validate the dev-cluster scenario

File: templates/flavors/windows-containerd/kustomization.yaml
Comment (jsturtevant) [inline templates/flavors/windows-containerd/kustomization.yaml:5]: We have not been installing Cluster Resource Sets for the default templates, only for the E2E tests.  This is mostly because cluster api doesn't include those definitions in what is considered a cluster.

 We shouldn't include this in the default template as all addon and other components (such as cni) needs to be managed by the user after the base cluster is up and running.

Since CAPI doesn't deploy CNI's if a customer used this template with a CRS I think they would get errors since they `clusterctl generate cluster <template>` doesn't include a CNI.  I think CRS are also an experimental feature?

The way we are handling this as of today with tilt is to deploy during tilt initialization like:

https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/e2abea2a0139ff9f8fd5e55dccc31a06208088e2/Tiltfile#L223-L225

Comment (marosset) [reply templates/flavors/windows-containerd/kustomization.yaml]: Thanks, I was able to get things working this way for dev clusters!

**Comment by marosset** [comment]
I was able to manually validate that the `windows-containerd` and `machinepool-windows-containerd` workload clusters had csi-proxy running when deployed through tilt (and not deployed for `windows` or `machinepool-windows` as expected.
/hold cancel

**Comment by jsturtevant** [comment]
We aren't running the tests against the csi proxy but they are modifying the templates to add them into the CI cluster templates so we should those tests to make sure it doesn't cause any issues

/test pull-cluster-api-provider-azure-windows-containerd-upstream-with-ci-artifacts
/test pull-cluster-api-provider-azure-conformance

**Comment by marosset** [comment]
It looks like the azure-disk/azure-file csi jobs are targeting `cluster-template-prow.yaml` based on 
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_azuredisk-csi-driver/1133/pull-kubernetes-e2e-capz-azure-disk-windows/1481582126804504576/build-log.txt

# Create workload Cluster.
/home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/hack/tools/bin/envsubst-v2.0.0-20210730161058-179042472c46 < /home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/templates/test/ci/cluster-template-prow.yaml

Let me update this PR so csi-proxy is installed for those cluster templates

**Comment by marosset** [comment]
> It looks like the azure-disk/azure-file csi jobs are targeting `cluster-template-prow.yaml` based on https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_azuredisk-csi-driver/1133/pull-kubernetes-e2e-capz-azure-disk-windows/1481582126804504576/build-log.txt
> 
> # Create workload Cluster.
> /home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/hack/tools/bin/envsubst-v2.0.0-20210730161058-179042472c46 < /home/prow/go/src/sigs.k8s.io/cluster-api-provider-azure/templates/test/ci/cluster-template-prow.yaml
> 
> Let me update this PR so csi-proxy is installed for those cluster templates

I made these updates

**Comment by marosset** [comment]
/test pull-cluster-api-provider-azure-ci-entrypoint
Re-triggering test to pick up https://github.com/kubernetes/test-infra/pull/25082

File: Tiltfile
Comment (CecileRobertMichon) [inline Tiltfile:220]: unrelated to this PR but @jsturtevant do you know why we are deleting the flannel-windows-addon cm but not creating it?

Comment (jsturtevant) [reply Tiltfile:220]: it's created on line https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2012/files#diff-c2ee8653e1d6b85f0aadf87cd438a9250806c052877248442be4d434cbc52425R226

thought it doesn't look like we delete `calico-windows-addon` which is added the line after.  Not sure if deletion is required... Seems to be working, even on reboots of tilt

Comment (marosset) [reply Tiltfile:220]: I was going to open another PR to fix that

**Comment by marosset** [comment]
I rebased after #2043 merged and we can now see that csi-proxy container is running on the windows nodes

https://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/2012/pull-cluster-api-provider-azure-ci-entrypoint/1491093865342439424/artifacts/workload-cluster/capz-d51w-kt8hv/kube-system_csi-proxy-xkbt2_56e72841-a720-4539-8b70-b1a1981916a6/

@jsturtevant @mboersma can you all take another look at this PR?

Thanks!

**Comment by jsturtevant** [comment]
@marosset could you squash the commits? lgtm but I would also like to run the conformance tests since those templates are modified though I don't expect anything.

**Comment by marosset** [comment]
> @marosset could you squash the commits? lgtm but I would also like to run the conformance tests since those templates are modified though I don't expect anything.

I added the label to have tide squash the commits for me. Is that sufficient?

**Comment by jsturtevant** [comment]
This has come [up in past](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1621#issuecomment-902804903). PR release notes don't show up in the release notes with this label.

**Comment by marosset** [comment]
> This has come [up in past](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1621#issuecomment-902804903). PR release notes don't show up in the release notes with this label.

Ah ok.
I'll rebase shortly.

**Review by devigned — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
